presto-main/src/main/java/io/prestosql/statestore/StateFetcher.java

```diff
@@ -274,6 +288,23 @@ private void handleExpiredQueryState(SharedQueryState state)
                         ((StateMap) stateCollection).put(newState.getBasicQueryInfo().getQueryId().getId(), stateJson);
                     }
                 }
+
+                // update query to failed in QUERY_STATE_COLLECTION_NAME if exists
+                stateCollection = stateStore.getStateCollection(QUERY_STATE_COLLECTION_NAME);
+                StateCollection finishStateCollection = stateStore.getStateCollection(FINISHED_QUERY_STATE_COLLECTION_NAME);
+                if (stateCollection != null && stateCollection.getType().equals(StateCollection.Type.MAP)) {
+                    Map<String, String> queryStateMap = ((StateMap<String, String>) stateCollection).getAll();
+                    if (queryStateMap.get(state.getBasicQueryInfo().getQueryId().getId()) != null) {
+                        killQuery(state.getBasicQueryInfo().getSelf().getHost(), state.getBasicQueryInfo().getSelf().getPort(), state.getBasicQueryInfo().getQueryId().getId());
```

Comment:
I don't think we should kill the query in StateFetcher and it's not a good idea to start a httpClient here.

Reply:
KillQuery is in case the coordinator not crash, just hang on for a while because of full GC or other reasons. 
If no killQuery, maybe the expired state has been cleaned, but query is also running if coordinator just hang on?

Reply:
Yea in such case it's really hard to tell if the coordinator is gone or just not responding in time. I think maybe it's better to kill the expired query in the `SqlQueryManager` instead of here(similar to the logic of killing OOM queries), and we also need to make sure the expiration config is long enough so we won't kill running query too frequently.

Reply:
Also in line 296 I don't think we need to fetch all the states we can use `StateMap.get()` instead

Reply:
Yes, The expiration config needs long enough indeed. 
I don't understand how to killQuery in SqlQueryManager like OOM Query killer. SqlQueryManager kills OOM query by using QueryExecution.fail(), but the expired query only has SharedQueryState(because the QueryExecution only exists in memory of the bad Coordinator) which cannot be killed like QueryExecution directly. I think the way to kill expired query is to post a killing request.

Reply:
Well.....I think twice about the way to kill expired query. Only the coordinator which not responding in time needs kill the expired query, so your suggestion is right. It's better to kill expired query in SqlQueryManager. I will do a try. Thanks!